design/dpop.md

```diff
@@ -0,0 +1,259 @@
+* Token Introspection
+* Token Revocation
+
+The following endpoints use cookie authentication:
+
+* Logout  
+* Check Session iframe   
+
+Therefore, no other endpoints should be affected.
+
+### Models
+
+#### Access Token
+
+New property, `jkt`, should be recognized as a member of the `cnf` claim.
+
+Classes/methods affected:
+
+* org.keycloak.representations.AccessToken.CertConf
+	* rename to Conf or Confirmation; introduce `jwkThumbprint` property
+        
+
+#### Client
+
+Properties should be added to the client model to support the following DPoP-related settings:
+
+* DPoP mode (enabled/optional/disabled)
+* DPoP proof lifetime (integer)
+* DPoP clock skew (integer)
+
+Classes/methods affected:
+
+* org.keycloak.models.ClientModel
+* org.keycloak.models.cache.infinispan.ClientAdapter
+* org.keycloak.models.cache.infinispan.entities.CachedClient
+* org.keycloak.models.jpa.ClientAdapter
+* org.keycloak.models.jpa.entities.ClientEntity
+* org.keycloak.models.map.client.AbstractClientEntity
+
+### Adapters
+
+#### Java Adapters
+
+In the Bearer only mode, Java adapters must:
+
+* enforce DPoP proof validation in the “Enabled/Required” mode;  
+* allow for DPoP validation of DPoP-bound tokens in the “Optional” mode;
+* ignore DPoP proofs in the “Disabled” mode.
+    
+The following options should be exposed in the adapter configuration:
+
+* DPoP Mode: required/optional/disabled
+* DPoP Proof Lifetime
+* DPoP Allowed Clock Skew
+
+> :question: Should we support DPoP for the “traditional” (monolithic) web applications, where adapters themselves handle Authorization Code flow, and the tokens are not exposed to the user agent?
+
+Classes/methods affected:
+
+*   org.keycloak.adapters.\*
+    
+
+##### Application Clustering
+
+As per the spec,
+
+> Servers SHOULD store, in the context of the request URI, the `jti` value of each DPoP proof for the time window in which the respective DPoP proof JWT would be accepted and decline HTTP requests to the same URI for which the `jti` value has been seen before.
+
+> :question: How do we implement DPoP proof replay protection in a clustered environment?  
+> 
+> Currently, only SAML adapters for Wildfly/AS/JBossWeb support distributed sessions via Infinispan.
+
+#### JavaScript Adapter
+
+In addition to DPoP handling proper, JavaScript adapter should also expose a method that would allow developers to obtain the DPoP proof header, since the latter would be required to access protected resources. As DPoP proof is bound to a particular HTTP method and URL, it should be recomputed for every protected resource request.
+
+Authors of [axios-keycloak][12], [keycloak-angular][13] and similar HTTP interceptors should be notified so that DPoP support could be introduced in the respective projects.
+
+Files affected:
+
+*   adapters/oidc/js/src/main/resources/keycloak.d.ts
+    
+
+#### Other Adapters
+
+Support in Node.JS adapter, mod\_authn\_oidc and Keycloak Gatekeeper could be also considered.
+
```

Comment:
[As you may know](https://www.keycloak.org/2020/08/sunsetting-louketo-project.adoc), Keycloak Gatekeeper has already been EOL.

Reply:
Oh indeed. Though OAuth2 Proxy is meant to be the successor, I'm not sure if DPoP would be relevant for that type of adapters, so will remove the mentioning for now.